staging/src/k8s.io/client-go/scale/client.go

```diff
@@ -0,0 +1,205 @@
+func NewForConfig(cfg *restclient.Config, mapper meta.RESTMapper, resolver dynamic.APIPathResolverFunc, scaleKindResolver ScaleKindResolver) ScalesGetter {
+	return &scaleClient{
+		config: cfg,
+		mapper: mapper,
+
+		clients:             make(map[schema.GroupVersion]restclient.Interface),
+		apiPathResolverFunc: resolver,
+		scaleKindResolver:   scaleKindResolver,
+		clientProvider: func(c *restclient.Config) (restclient.Interface, error) {
+			return restclient.RESTClientFor(c)
+		},
+	}
+}
+
+// clientAndMappingFor returns a cached rest client and the associated resource for the given GroupKind,
+// populating the cache if necessary.
+func (c *scaleClient) clientAndMappingFor(kind schema.GroupKind) (restclient.Interface, *meta.RESTMapping, error) {
+	c.clientsMu.Lock()
+	defer c.clientsMu.Unlock()
+
+	mapping, err := c.mapper.RESTMapping(kind)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to get rest mapping for %s: %v", kind.String(), err)
+	}
+
+	groupVer := mapping.GroupVersionKind.GroupVersion()
+
+	// check if we have an existing client
+	if existingClient, found := c.clients[groupVer]; found {
+		return existingClient, mapping, nil
+	}
+
+	// avoid changing the original config
+	confCopy := *c.config
+	conf := &confCopy
+
+	// we need to set the API path based on GroupVersion (defaulting to the legacy path if none is set)
+	conf.APIPath = c.apiPathResolverFunc(mapping.GroupVersionKind)
+	if conf.APIPath == "" {
+		conf.APIPath = "/api"
+	}
+	// NB: the GV here is used to set the API path, but has little relevance to the
+	// serialized group-versions
+	conf.GroupVersion = &groupVer
+	conf.NegotiatedSerializer = serializer.DirectCodecFactory{
+		CodecFactory: codecs,
+	}
+
+	if len(conf.UserAgent) == 0 {
+		conf.UserAgent = restclient.DefaultKubernetesUserAgent()
+	}
+
+	cl, err := c.clientProvider(conf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.clients[groupVer] = cl
+
+	return cl, mapping, nil
+}
+
+// namespacedScaleClient is an ScaleInterface for fetching
+// Scales in a given namespace.
+type namespacedScaleClient struct {
+	client    *scaleClient
+	namespace string
+}
+
+func (c *scaleClient) Scales(namespace string) ScaleInterface {
+	return &namespacedScaleClient{
+		client:    c,
+		namespace: namespace,
+	}
+}
+
+func (c *namespacedScaleClient) Get(kind schema.GroupKind, name string) (*autoscaling.Scale, error) {
+	// Currently, a /scale endpoint can return different scale types.
+	// Until we have support for the alternative API representations proposal,
+	// we need to deal with accepting different API versions.
+	// In practice, this is autoscaling/v1.Scale and extensions/v1beta1.Scale
+
+	client, mapping, err := c.client.clientAndMappingFor(kind)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get client for %s: %v", kind.String(), err)
+	}
+
+	rawObj, err := client.Get().
+		Namespace(c.namespace).
+		Resource(mapping.Resource).
+		Name(name).
+		SubResource("scale").
+		Do().
```

Comment:
If you switched to use a `DoRaw`, then used your custom scheme for decoding, wouldn't you be able to use any given `rest.Interface` without constructing a new client from config?  I like supporting `NewForConfig`, I'm just not seeing why I need to have multiple clients and not supporting `New`.

Reply:
We still need a `rest.Interface` for each group-version, no?

Reply:
> We still need a rest.Interface for each group-version, no?

Is it because you need to point to different endpoints? You can construct a path and call `AbsPath()`. We can expose the function the restclient uses to build the path so that we don't have to hard code here.

Reply:
Actually making `defaultServerURLFor` public seems reasonable, but just circumventing the request with `AbsPath` seems hacky.  I think if we're going to change this, we should have a type that provides clients for a given GroupVersion only, and use it in both the scale client and dynamic client (since the dynamic client currently uses the same approach that I have here).

For example:

```go
type Provider interface {
    GroupVersion(gv schema.GroupVersion) rest.Interface
}

provider.GroupVersion(schema.GroupVersion{"apps", "v1beta2"}).Get()
  .Resource("deployments")
  .Name("foo")
  .SubResource("scale")
  .Do()
```

That looks a bit clunky with the `GroupVersion` before the Get, but provides a rest.Interface out of it, so it fits with more existing code.  An alternative might be:

```go
type MultiVersionInterface interface {
    // the same methods as rest.Interface, but they return a GroupVersionProvider below
}

type GroupVersionProvider interface {
    GroupVersion(schema.GroupVersion) rest.Request
}

type client MultiVersionInterface = ...
client.Get().GroupVersion(schema.GroupVersion{"apps", "v1beta2"})
   .Resouce("deployments")
   .Name("foo")
   .Do()
```

But since that doesn't return a `rest.Interface` anywhere, it's mostly useful only here and in the dynamic client.

WDYT?

Reply:
> But since that doesn't return a rest.Interface anywhere, it's mostly useful only here and in the dynamic client.
> 
> WDYT?

I think the dynamic client interface needs some work, but I would definitely accept an `.AbsPath` here to make the externally facing bits of this API easier to work with.

Reply:
Let's use `.AbsPath` here and refactor it with the dynamic client in another PR.

Also, with the first approach, it looks like the provider still returns different restclient for different group versions. That sounds unnecessary.

Before taking the second approach, could we try to add a method to the rest.Request to allow us change group/version after its construction?